flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/keyselector/AttributeBasedJoinKeyExtractor.java

```diff
@@ -0,0 +1,299 @@
+     *
+     * @param joinAttributeMap Map defining equi-join conditions. Outer key: inputId (>= 1). Inner
+     *     key: {@link AttributeRef} to a field in a *previous* input. Inner value: {@link
+     *     AttributeRef} to the corresponding field in the *current* input (inputId == outer key).
+     * @param inputTypes Type information for all input streams (indexed 0 to N-1).
+     */
+    public AttributeBasedJoinKeyExtractor(
+            final Map<Integer, Map<AttributeRef, AttributeRef>> joinAttributeMap,
+            final List<InternalTypeInfo<RowData>> inputTypes) {
+        this.joinAttributeMap = joinAttributeMap;
+        this.inputTypes = inputTypes;
+    }
+
+    @Override
+    public RowData getKeyForStateStorage(RowData row, int inputId) {
+        if (inputId == 0) {
+            // Input 0 uses the fixed default key as it's the start of the join chain.
+            return DEFAULT_KEY;
+        }
+
+        // For inputs > 0, storage key derived from current row's equi-join fields.
+        final Map<AttributeRef, AttributeRef> attributeMapping = joinAttributeMap.get(inputId);
+        if (attributeMapping == null || attributeMapping.isEmpty()) {
+            // No equi-join conditions defined for this input, use default key.
+            return DEFAULT_KEY;
+        }
+
+        // Indices of fields in the *current* input (inputId) used as the *right* side of joins.
+        final List<Integer> keyFieldIndices = determineKeyFieldIndices(inputId);
+
+        if (keyFieldIndices.isEmpty()) {
+            // Mappings exist, but none point to fields *within* this inputId (config error?), use
+            // default key.
+            return DEFAULT_KEY;
+        }
+
+        return buildKeyRow(row, inputId, keyFieldIndices);
+    }
+
+    @Override
+    public RowData getKeyForStateLookup(int depth, RowData[] currentRows) {
+        if (depth == 0) {
+            // Input 0 lookup always uses the fixed default key.
+            return DEFAULT_KEY;
+        }
+
+        // For depths > 0, lookup key derived from *previous* rows (indices < depth)
+        // using the *left* side of equi-join conditions for the *current* depth.
+        final Map<AttributeRef, AttributeRef> attributeMapping = joinAttributeMap.get(depth);
+        if (attributeMapping == null || attributeMapping.isEmpty()) {
+            // No equi-join conditions link previous inputs to this depth (e.g. cross join).
+            // Use default key.
+            return DEFAULT_KEY;
+        }
+
+        // TreeMap ensures deterministic key structure: left inputId -> left fieldIndex
+        final Map<Integer, Map<Integer, Object>> sortedKeyComponents = new TreeMap<>();
+
+        // Iterate through join attributes for the current depth.
+        // Key (leftAttrRef) points to previous input (< depth).
+        // Value (rightAttrRef) points to current input (== depth).
+        for (Map.Entry<AttributeRef, AttributeRef> entry : attributeMapping.entrySet()) {
```

Comment:
Why do we need to iterate over all the entries? Can't we precompute indices for a given join depth?
I'd imagine in a pseudocode:
```
val mask = joinMasks.get(depth);
RowData value = currentRows[depth];
RowData key = mask.apply(value);
```

I am a bit afraid of doing all the expensive computations for every record.

Reply:
That's a valid concern. It's not so straightforward as you have in your pseudocode because we build the index/key based on all current rows < depth. Basically, we build the key based on the left side of the join before iterating through the rows of depth, "the right side".

Mostly, for join conditions with unique key, we'll extract the key only for one record. However, I agree we had unnecessary expensive recomputations here in case we do not have primary keys. I refactored the code so we precompute these extractors only once and use them for all records 🙂 See: https://github.com/apache/flink/pull/26313/commits/0aa71e40ebd4120bf17d3a19e4acbb39217626f6